### ISSUE TITLE
Strip packaged binaries to reduce buildpack size

### ISSUE THREAD

**Issue by edmorley** [heroku/libcnb.rs#319]
Stripping the packaged binaries will reduce the buildpack size, which will help with builder image pull times, and also local development/integration test runtimes (the copy of local buildpacks into the ephemeral builder image when using `--buildpack` is surprisingly slow).

Rust 1.58 stabilised the `-C strip` option:
https://doc.rust-lang.org/stable/rustc/codegen-options/index.html#strip

However it's not until Rust 1.59 (currently in beta, due for release 2022-02-24) that corresponding support will be added to Cargo:
https://doc.rust-lang.org/beta/cargo/reference/profiles.html#strip

The feature can be enabled by adding this to `Cargo.toml` for the crate in question:

```toml
[profile.release]
strip = true
```

However this would mean every project using libcnb would need to remember to do this, so I think it would make more sense if we configure it automatically in `libcnb-cargo`.

I think we may want to enable it for non-release libcnb-cargo builds too, since it will speed up development use-cases as mentioned above. (And debug symbols don't exactly help much when the buildpack is run inside the container as part of the lifecycle, since no easy way to run via a debugger anyway.)

Profile options like this can't be passed via `cargo build` arguments:
https://doc.rust-lang.org/beta/cargo/commands/cargo-build.html

...but we should be able to enable it via env var:
https://doc.rust-lang.org/beta/cargo/reference/config.html#environment-variables

For example:

```
CARGO_PROFILE_RELEASE_STRIP=true
CARGO_PROFILE_DEV_STRIP=true
```

GUS-W-11324847.

**Comment by edmorley** [comment]
Experimenting with this using Rust 1.59 beta, by running `CARGO_PROFILE_RELEASE_STRIP=true cargo libcnb package --release`, I get the following size reductions:

- `example-01-basics`: 12.0 MB -> 4.1 MB
- `example-02-ruby-sample`: 20.5 MB -> 8.5 MB

And for non-release builds (using `CARGO_PROFILE_DEV_STRIP=true cargo libcnb package`), I get:

- `example-01-basics`: 50.4 MB -> 10.4 MB
- `example-02-ruby`: 97.6 MB -> 20.7 MB

During the build, this warning appears:

```
   Compiling libcnb-proc-macros v0.1.1 (/Users/emorley/src/libcnb.rs/libcnb-proc-macros)
warning: stripping debug info with `strip` failed: exit status: 1
  |
  = note: /Library/Developer/CommandLineTools/usr/bin/strip: error: symbols referenced by indirect symbol table entries that can't be stripped in: /Users/emorley/src/libcnb.rs/target/release/deps/liblibcnb_proc_macros-f34f391f2c4d9c8e.dylib
          __NSGetArgc
          __NSGetArgv
          __NSGetEnviron
          __NSGetExecutablePath
          __Unwind_Backtrace
          __Unwind_DeleteException
          __Unwind_GetCFA
          __Unwind_GetDataRelBase
          __Unwind_GetIP
          __Unwind_GetIPInfo
          __Unwind_GetLanguageSpecificData
          __Unwind_GetRegionStart
          __Unwind_GetTextRelBase
          __Unwind_RaiseException
          __Unwind_Resume
          __Unwind_SetGR
          __Unwind_SetIP
...
```

The build completes successfully however.

What's interesting about this, is that the path referenced in the error message is under `target/release/` not `target/x86_64-unknown-linux-musl/release/`. This plus the fact it's for the `libcnb-proc-macros` crate, makes me think it's the stripping of the compile-time regex validation that's failing (which is compiled for the host), rather than the cross-compiled crate itself.

Trying again using a strip mode of `debuginfo` rather than `true` (`true` being equivalent to `symbols`, the maximum level), the warning goes away, and I get the following (slightly larger) output sizes:

- `example-01-basics`: 5.4 MB (release), 14.5 MB (dev)
- `example-02-ruby`: 11.3 MB (release), 28.9 MB (dev)

The Cargo docs mention a `build-override` setting:
https://doc.rust-lang.org/beta/cargo/reference/config.html#profilenamebuild-override

> The build-override table overrides settings for build scripts, proc macros, and their dependencies. It has the same keys as a normal profile.

We can take advantage of that by using:
`CARGO_PROFILE_DEV_STRIP=true CARGO_PROFILE_DEV_BUILD_OVERRIDE=debuginfo cargo libcnb package`

...which prevents the warning, whilst still giving us the full output size reductions.

It seems like perhaps this might be a Rust bug where it's not fully taking into account the strip options required for the host vs the target, but either way, there is an easy workaround for now :-)

**Comment by edmorley** [comment]
> It seems like perhaps this might be a Rust bug

Filed:
https://github.com/rust-lang/rust/issues/93988

**Comment by edmorley** [comment]
Now available in stable:
https://blog.rust-lang.org/2022/02/24/Rust-1.59.0.html#creating-stripped-binaries

**Comment by edmorley** [comment]
> Filed:
> [rust-lang/rust#93988](https://github.com/rust-lang/rust/issues/93988)

The warnings seen when stripping using local path dependencies (ie when hacking on libcnb in this repo) will now be resolved in Rust 1.63, thanks to:
https://github.com/rust-lang/rust/pull/97991

That needn't block the work on enabling stripping in libcnb, since (a) they are only warnings, (b) they aren't shown when people use libcnb from crates.io (only local path based builds).
